src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -1354,13 +1357,156 @@ public ValidateCountAttribute(int minLength, int maxLength)
     }
 
     /// <summary>
+    /// Base class for all valid values generators with cache support.
+    /// </summary>
+    public abstract class CachedValidValuesGeneratorBase : IValidateSetValuesGenerator
+    {
+        // Cached valid values.
+        private string[] _validValues;
+
+        /// <summary>
+        /// Sets a time interval in seconds to reset the '_validValues' dynamic valid values cache.
+        /// By default 'ValidValuesCacheExpiration = 0' and no valid values is cached (is re-evaluated every time it is used).
+        /// </summary>
+        public int ValidValuesCacheExpiration { get; set; } = 0;
+
+        /// <summary>
+        /// Abstract method to generate a valid values.
+        /// </summary>
+        public abstract string[] GenerateValidValues();
+
+        /// <summary>
+        /// Get a valid values.
+        /// </summary>
+        public IEnumerable<string> GetValidValues()
+        {
+            if (_validValues != null)
+            {
+                return _validValues;
+            }
+
+            var _validValuesNoCache = GenerateValidValues();
+
+            if (_validValuesNoCache == null)
+            {
+                throw new ValidationMetadataException(
+                    "ValidateSetGeneratedValidValuesListIsEmpty",
+                    null,
+                    Metadata.ValidateSetGeneratedValidValuesListIsEmpty);
+            }
+
+            if (ValidValuesCacheExpiration > 0)
+            {
+                Task.Delay(ValidValuesCacheExpiration).ContinueWith((task) => _validValues = null);
+                _validValues = _validValuesNoCache;
+            }
+
+            return _validValuesNoCache;
+        }
+    }
+
+    /// <summary>
     /// Validates that each parameter argument is present in a specified set
     /// </summary>
     [AttributeUsage(AttributeTargets.Field | AttributeTargets.Property)]
     public sealed class ValidateSetAttribute : ValidateEnumeratedArgumentsAttribute
     {
+        // We can use either static '_validValues'
+        // or dynamic valid values list generated by instance of 'validValuesGeneratorType'.
         private string[] _validValues;
 
+        // The valid values generator cache works across 'ValidateSetAttribute' instances.
+        private static ConcurrentDictionary<Type, ValidValuesGeneratorCacheEntry> s_ValidValuesGeneratorCache = new ConcurrentDictionary<Type, ValidValuesGeneratorCacheEntry>();
```

Comment:
>ValidValuesGeneratorCacheEntry [](start = 50, length = 30)

I think this can be `IValidateSetValuesGenerator`.
There will be very few types that implement this interface, so I think the extra complexity around cache cleanup on the generator is not needed.